client/components/modal.jsx

```diff
@@ -0,0 +1,148 @@
+import React from 'react';
+import Search from './search-bar';
+
+export default class TrainingModal extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      userId: 1,
+      exercises: [],
+      filteredExercises: [],
+      showSuggestions: false,
+      userInput: '',
+      addExercise: [],
+      exerciseId: this.props.exerciseId,
+      ModalisOpen: this.props.ModalisOpen,
+      newSet: this.props.newSet,
+      date: this.props.date
+    };
+    this.onType = this.onType.bind(this);
+    this.onClick = this.onClick.bind(this);
+    this.addExercise = this.addExercise.bind(this);
+    this.handleChange = this.handleChange.bind(this);
+    this.addSet = this.addSet.bind(this);
+  }
+
+  componentDidMount() {
+    fetch('/api/exercise-list')
+      .then(res => res.json())
+      .then(exercises => this.setState({ exercises }));
+  }
+
+  addExercise() {
+    event.preventDefault();
+    for (let i = 0; i < this.state.exercises.length; i++) {
+      if (this.state.userInput === this.state.exercises[i].exercise) {
+        this.setState({
+          exerciseId: this.state.exercises[i].exerciseId
+        });
+      }
+    }
+    this.setState({
+      addExercise: this.state.userInput,
+      userInput: ''
+    });
+  }
+
+  addSet() {
+    event.preventDefault();
+    this.setState(({
+      newSet: [...this.state.newSet, { reps: '', weight: '' }]
+    }));
+  }
+
+  removeSet(index) {
+    const newSet = this.state.newSet;
+    newSet.splice(index, 1);
+    this.setState({ newSet });
+  }
+
+  handleChange(index, event) {
+    const newSet = this.state.newSet;
```

Comment:
same here. this is a little more involved because you want to mutate neither the array, nor the object to replace.
```js
const setCopy = this.state.newSet.slice(0);
const exerciseCopy = Object.assign({}, setCopy[index])
exerciseCopy[event.target.name] = event.target.value
setCopy[index] = exerciseCopy
this.setState({ newSet: setCopy })
```